### PR TITLE
add project dependency packaging

### PR DESCRIPTION
**Pull request by andz-bb**
## Description

adds the dependency and package foundation for projects.

- discovers dependencies from structured references without treating ordinary text as a dependency
- packages assigned resources with their owned content, preserving deliberate exclusions
- validates packages and remaps resource, view and row-action references into the destination workspace
- removes credentials and records setup requirements for imported resources
- bounds archive extraction and encrypted expansion, locks package operations and rolls back incomplete imports

covers owner exclusions, malformed packages, credential removal, remapping and import cleanup with API tests.

part 1 of the six-part projects stack. review against master. next: [#19245](https://github.com/Budibase/budibase/pull/19245).

## Addresses

- https://github.com/Budibase/budibase/issues/18349

## App Export

- n/a

## Screenshots

- n/a

## Launchcontrol

improves project export and import reliability, including portable views and credential removal

### CODE REVIEW

File: packages/server/src/sdk/workspace/resources/references.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+export interface ResourceSearchTarget extends UsedResource {
+  idToSearch: string
+  extraDependencies?: UsedResource[]
+  matchInBindings?: boolean
+  matchInEnabledTools?: boolean
+  matchExactly?: boolean
+}
+
+const STRUCTURED_REFERENCE_PROPERTIES = new Set(["_id", "dependencies", "id"])
```
[the remark this answers is not in the document]
Comment (andz-bb) [reply packages/server/src/sdk/workspace/resources/references.ts:12]: leaving this as-is: cloning returns the complete resource graph, including supporting resources that are not assignable roots. Project consumers apply the narrower root filter when needed, so removing _id here would break the shared clone and package contract.

File: packages/server/src/sdk/workspace/projects/crud.ts [hunk 2]
```diff
@@ -143,13 +136,7 @@ async function clearAssignments(projectId: string) {
...
 
   const assignmentUpdates = assignedDocs
-    .filter(
-      doc =>
-        doc._id &&
-        !isExternalTableID(doc._id) &&
-        !doc._id.startsWith(prefixed(DocumentType.DATASOURCE)) &&
-        hasProject(doc, projectId)
-    )
+    .filter(doc => doc._id && hasProject(doc, projectId))
```
[the remark this answers is not in the document]
Comment (andz-bb) [reply packages/server/src/sdk/workspace/projects/crud.ts:139]: leaving this as-is: the developer-preview data model does not need backwards compatibility. datasource ownership is canonical, and stripping nested projectIds avoids two competing assignment sources.

**Comment by andz-bb** [comment]
@cubic-dev-ai rerun

**Comment by andz-bb** [comment]
@cubic-dev-ai rerun

**Comment by andz-bb** [comment]
@cubic-dev-ai rerun

**Comment by andz-bb** [comment]
@cubic-dev-ai rerun

File: packages/server/src/sdk/workspace/projects/backups/imports.ts [hunk 3]
```diff
@@ -909,6 +1186,36 @@ const bulkInsertDocs = async (
...
+    )
+  )
+  insertedDocs.push({
+    _id: webhook._id!,
+    _rev: webhook._rev!,
+  })
+
+  trigger.webhookId = webhook._id
+  trigger.inputs = {
+    schemaUrl: `api/webhooks/schema/${workspaceId}/${webhook._id}/${webhook.schemaToken}`,
```
[the remark this answers is not in the document]
Comment (andz-bb) [reply packages/server/src/sdk/workspace/projects/backups/imports.ts:1214]: confirmed that a builder could send the import to a production workspace. changing schemaUrl alone would still fail because the new webhook exists only in that production database. fixed in 314be7b015 by applying the existing AppType.DEV request middleware, matching automation authoring endpoints. the new API regression submits a valid package to a published workspace and verifies rejection with no imported project; removing the middleware makes the test fail (200 instead of 400). no duplicate SDK guard or URL rewrite added.

File: packages/server/src/api/routes/tests/project.spec.ts [hunk 4]
```diff
@@ -1264,27 +1570,31 @@ describe("/projects", () => {
...
+      const datasource = await config.api.datasource.create({
+        ...basicDatasource().datasource,
+        projectIds: [project._id],
+      })
       const externalTableId = buildExternalTableId(datasource._id!, "TestTable")
       const externalTable = basicTable(datasource, {
         _id: externalTableId,
         name: "TestTable",
         primaryDisplay: `{{ ${externalTableId}.name }}`,
-        projectIds: [project._id],
```
[the remark this answers is not in the document]
Comment (andz-bb) [reply packages/server/src/api/routes/tests/project.spec.ts:1280]: agreed that these assertions were vacuous. datasource update and project import already strip entity memberships, and project deletion clears the assignable datasource root. removed the dead assertions and unused entity setup in dbe9ab9c62, including the same redundant setup in the multi-project deletion case. kept the root-membership and remaining-project checks. all 73 project tests pass.

**Comment by andz-bb** [comment]
@cubic-dev-ai rerun

File: packages/server/src/sdk/workspace/projects/backups/exports.ts [hunk 5]
```diff
@@ -94,55 +91,66 @@ export async function listAssignedAgentFiles(
...
-    doc: { _id?: string; name?: string },
-    type: ResourceType
-  ): UsedResource => ({
-    id: doc._id!,
-    name: doc.name || "Unknown",
-    type,
-  })
+  return sortResources(
+    assignedDocs.flatMap(doc => {
+      if (
```
[the remark this answers is not in the document]
Comment (andz-bb) [reply packages/server/src/sdk/workspace/projects/backups/exports.ts:96]: this stack deliberately keeps project memberships on owner resources: external tables and queries follow their datasource. the member index, dependency graph and save/assignment APIs use that same rule, and the PR description states it explicitly. the root `hasProject` requirement was already present at `dbe9ab9c`, before this cleanup; the new classifier preserves it.

`master` did include the entity-only fallback. restoring it only in the exporter would make its direct-member index disagree with the graph. migration compatibility for entity-only assignments is outside this draft stack's agreed scope; if that becomes a rollout requirement, it needs a consistent migration across these boundaries. keeping the existing owner-membership contract here.

File: packages/server/src/api/routes/tests/resource.spec.ts [hunk 6]
```diff
@@ -986,33 +1121,36 @@ describe("/api/resources/usage", () => {
       })
     })
 
-    it("duplicates project external tables through their datasource", async () => {
+    it("duplicates project datasources and strips external table project assignments", async () => {
```
[the remark this answers is not in the document]
Comment (andz-bb) [reply packages/server/src/api/routes/tests/resource.spec.ts:1124]: the retained test now seeds `entities[*].projectIds` directly in storage. the removed test supplied them through datasource update, which strips them during setup, so its final assertion did not exercise copying.

nested entity memberships are stripped unconditionally in the shared copy path, regardless of whether the project is copied. project availability only affects the separate owner-membership filter, whose no-project and existing-project cases remain covered. i temporarily removed the nested stripping statement and verified that the retained test failed at its final assertion, then restored the production file byte-for-byte and reran it successfully. keeping this smaller fixture with meaningful coverage.

File: packages/builder/src/components/integration/APIEndpointViewer.svelte [hunk 7]
```diff
@@ -335,14 +322,18 @@
     ? prettifyQueryRequestBody(editableQuery, mergedBindings)
     : undefined
 
+  const stripQueryProjectIds = (query: Query): Query => {
```
Comment (melohagan) [inline packages/builder/src/components/integration/APIEndpointViewer.svelte:325]: Personally I'd prefer to keep the code simple and not worry about this backwards compatibility issue. Is there much harm done if projectIds are left in some existing queries?

File: packages/server/src/ai/tools/budibase/automations.ts [hunk 8]
```diff
@@ -161,11 +164,7 @@ const createAutomationTools = (
     )
     .map((automation): BudibaseToolDefinition => {
       const automationName = automation.name || automation._id!
-      const sanitizedAutomationId = automation._id!.replace(
-        /[^A-Za-z0-9_-]/g,
-        "_"
-      )
-      const toolName = `${sanitizedAutomationId}_trigger`.substring(0, 64)
```
Comment (melohagan) [inline packages/server/src/ai/tools/budibase/automations.ts:168]: Why is this changed in this PR?

File: packages/server/src/api/controllers/datasource.ts [hunk 9]
```diff
@@ -217,10 +217,7 @@ async function invalidateVariables(
 const isDatasourceEntity = (entity: unknown): entity is Table =>
   typeof entity === "object" && entity !== null && !Array.isArray(entity)
 
-const resolveDatasourceEntityProjectIds = async (
-  datasource: Datasource,
-  existingDatasource?: Datasource
-) => {
+const stripDatasourceEntityProjectIds = (datasource: Datasource) => {
```
Comment (melohagan) [inline packages/server/src/api/controllers/datasource.ts:220]: Similar to the APIEndpointViewer, my personal preference would be to exclude this code for simplicity.

File: packages/server/src/api/routes/tests/project.spec.ts [hunk 10]
```diff
@@ -139,6 +123,38 @@ describe("/projects", () => {
...
+    "removes uploaded files when import rejects %s",
+    async rejection => {
+      await withProjectsEnabled(async () => {
+        const remove = jest.spyOn(fsp, "rm")
+        try {
+          const request = config
+            .request!.post("/api/projects/import")
+            .set(config.defaultHeaders())
+            .attach("file", Buffer.from("first"), "first.tar.gz")
+          if (rejection === "invalid fields") {
```
Comment (melohagan) [inline packages/server/src/api/routes/tests/project.spec.ts:136]: Generally speaking it's better to avoid logic in tests, i.e. just have two tests instead. You can still create helper functions to share code if it makes the tests more readable.

File: packages/server/src/api/routes/tests/project.spec.ts [hunk 11]
```diff
@@ -139,6 +123,38 @@ describe("/projects", () => {
...
+            request.field("encryptPassword", "x".repeat(1025))
+          } else {
+            request.attach("file", Buffer.from("second"), "second.tar.gz")
+          }
+          await request.expect(400)
+
+          expect(remove).toHaveBeenCalledTimes(
+            rejection === "invalid fields" ? 1 : 2
```
Comment (melohagan) [inline packages/server/src/api/routes/tests/project.spec.ts:144]: Similarly better to just have two tests instead of doing this

File: packages/server/src/api/routes/tests/project.spec.ts [hunk 12]
```diff
@@ -845,20 +851,32 @@ describe("/projects", () => {
       })
     })
 
-    it("preserves query assignments", async () => {
```
Comment (melohagan) [inline packages/server/src/api/routes/tests/project.spec.ts:848]: I'd probably prefer keeping this as it was rather than having to handle the legacy edge case if possible.

File: packages/server/src/api/routes/tests/project.spec.ts [hunk 13]
```diff
@@ -1159,54 +1394,112 @@ describe("/projects", () => {
...
-              type: "agent_linked_content",
+              type: "excluded_dependency",
               count: 1,
             }),
           ])
         )
       })
     })
+
+    it("ignores resource ids in ordinary text", async () => {
```
Comment (melohagan) [inline packages/server/src/api/routes/tests/project.spec.ts:1461]: It's difficult to match the description of the test to the implementation and expectations.

File: packages/server/src/api/routes/tests/project.spec.ts [hunk 14]
```diff
@@ -1312,18 +1630,74 @@ describe("/projects", () => {
...
+            toolName: getRowToolNames(externalTableId).create_row,
+            executionPrincipal: ToolExecutionPrincipal.ADMIN,
+          },
+        ],
+        allowKnowledgeSourceDownload: true,
+      })
+
+      const resourceGraph = await config.api.resource.getResourceDependencies()
+      expect(
+        resourceGraph.body.resources[agent._id!].dependencies.map(
```
Comment (melohagan) [inline packages/server/src/api/routes/tests/project.spec.ts:1689]: It could help make the test a little more readable to put:

```ts
resourceGraph.body.resources[agent._id!].dependencies.map(
          dependency => dependency.id
        )
```

AND 

```ts
resourceGraph.body.resources[project._id].dependencies.map(
          dependency => dependency.id
        )
```

Into a variable. I'm struggling to see how this relates to the agent tools

File: packages/server/src/api/routes/tests/project.spec.ts [hunk 15]
```diff
@@ -1385,58 +1759,115 @@ describe("/projects", () => {
...
+                })
+              ).appId
+
+        await config.withHeaders(
+          { [Header.WORKSPACE_ID]: destinationWorkspaceId },
+          async () => {
+            const tableIds = [table._id!]
+            const viewIds = [view.id]
+            const rowActionIds = [rowAction.id]
+            for (let importIndex = 0; importIndex < 2; importIndex++) {
```
Comment (melohagan) [inline packages/server/src/api/routes/tests/project.spec.ts:1814]: This is tricky to understand for a test. Is there a way to make the intention clearer? Like initial import, and then re-import, or whatever is happening here 😅

File: packages/server/src/api/routes/tests/project.spec.ts [hunk 16]
```diff
@@ -1823,6 +2400,87 @@ describe("/projects", () => {
...
+    "rejects a missing %s resource",
+    async location => {
+      await withProjectsEnabled(async () => {
+        const missing = {
+          id: "au_missing",
+          name: "Missing automation",
+          type: "automation",
+        }
+        const dependencyIndex =
+          location === "dependency"
```
Comment (melohagan) [inline packages/server/src/api/routes/tests/project.spec.ts:2413]: Have two test cases instead of test logic

File: packages/server/src/api/routes/tests/project.spec.ts [hunk 17]
```diff
@@ -1823,6 +2400,87 @@ describe("/projects", () => {
...
+        const packageBuffer = await createTarPackage(
+          createMinimalPackageEntries({
+            manifest: {
+              resourcesByType: {
+                project: 1,
+                table: 1,
+              },
+            },
+            dependencyIndex: {
+              directMembers: location === "direct member" ? [mistyped] : [],
```
Comment (melohagan) [inline packages/server/src/api/routes/tests/project.spec.ts:2455]: Have two test cases instead of test logic.

**Review by melohagan — CHANGES_REQUESTED** [review]
I'm about a third of the way through the PR, will resume tomorrow. 

You're welcome to wait for the full review, or address the current comments. Whatever suits. 

General impression so far is:

1. A lot of the tests are very complex and difficult to read and understand. Helper functions with descriptive names could reduce the size of the test and make the intent more obvious.

2. There's some edge case handling that I think is more bloat than necessity, but you may disagree on that!

File: packages/server/src/sdk/workspace/datasources/datasources.ts [hunk 18]
```diff
@@ -40,6 +40,42 @@ import { getEnvironmentVariables } from "../../utils"
 import { ensureValidPrimaryDisplay } from "../tables/utils"
 
 const ENV_VAR_PREFIX = "env."
+const ENV_VAR_REFERENCE_PATTERN = new RegExp(
```
Comment (melohagan) [inline packages/server/src/sdk/workspace/datasources/datasources.ts]: NAB: It may be better to move some of this functionality into string-templates

File: packages/server/src/sdk/workspace/projects/backups/files.ts
Comment (melohagan) [inline packages/server/src/sdk/workspace/projects/backups/files.ts]: A security comment from Astra worth double checking.

```
One significant finding: directory entries bypass the resource limits, allowing an authenticated import to consume excessive disk space, inodes, and CPU.

  packages/server/src/sdk/workspace/projects/backups/files.ts:47 recurses into directories without incrementing either counter. Only files count towards the 2,000-file and byte limits. The validation before extraction has the same gap: it counts all entries but never limits that count. packages/server/src/sdk/workspace/
  projects/backups/imports.ts:213

  I reproduced this with 2,100 empty directories in a 15 KB archive. The walker accepted them and finished with { files: 0, bytes: 0 }. Much larger archives can create substantial filesystem work before the missing or invalid manifest is rejected. The depth limit does not restrict directory breadth.

  Fix: enforce a total entry limit, including directories, before extraction and during the filesystem walk. Keep the separate byte limit for file content.
```

File: packages/server/src/sdk/workspace/resources/index.ts [hunk 19]
```diff
@@ -12,40 +12,154 @@ import {
...
+      return []
+    }
+    return [
+      dependency,
+      ...collectTransitiveResourceDependencies(graph, dependency.id, seen),
+    ]
+  })
+}
+
+function collectCloneResourceDependencies(
```
Comment (melohagan) [inline packages/server/src/sdk/workspace/resources/index.ts]: I would consolidate this function with `collectTransitiveResourceDependencies`

> They perform the same graph walk and use the same seen set to deduplicate. The only behavior difference is a direct self reference

File: packages/server/src/sdk/workspace/resources/index.ts [hunk 20]
```diff
@@ -195,6 +376,13 @@ export async function getResourcesInfo(): Promise<
     }
   }
 
+  for (const datasource of datasources) {
+    searchForUsages(datasource._id!, {
```
Comment (melohagan) [inline packages/server/src/sdk/workspace/resources/index.ts]: Something like `addReferencedDependencies` might be a better function name as this function does not return search results, but instead calls `addDependency`

File: packages/server/src/sdk/workspace/resources/references.ts
Comment (melohagan) [inline packages/server/src/sdk/workspace/resources/references.ts]: A comment from Sol worth double checking:

```
I found two correctness issues:

  - Inactive bindings can create dependencies. packages/server/src/sdk/workspace/resources/references.ts:12 finds text that looks like a Handlebars block, then searches it without checking whether it is a comment or quoted literal. I verified that both {{! query_orders.rows }} and {{ "query_orders.rows" }} report query_orders. That can add an unrelated resource to the graph or report it as an excluded Project dependency. Comment handling is a clear fix; quoted content needs care because encoded JavaScript bindings can legitimately reference a resource inside a string.

  - A document’s own _id becomes a self dependency. The generic structured-field rule matches top-level _id at packages/server/src/sdk/workspace/resources/references.ts:129. This produces table → table and automation → automation edges. I would handle this at the caller, where the owner ID is known: a screen scanned on behalf of its app still needs its top-level _id matched.

  The code is also potentially expensive: it checks every target for every string value and builds regexes repeatedly for binding blocks.
```

**Review by melohagan — CHANGES_REQUESTED** [review]
Great work Andy! 

Looks good to me, just a couple of comments worth checking out.
